src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -102,9 +105,11 @@ static MemoryScope createShared(Object ref, Runnable cleanupAction) {
      * a confined scope and this method is called outside of the owner thread.
      */
     final void close() {
-        justClose();
-        if (cleanupAction != null) {
-            cleanupAction.run();
+        try {
+            justClose();
+            cleanupAction.cleanup();
```

Comment:
Won't explicitly calling close() end up calling disable() on BasicCleanupAction, and then when the cleaner calls close it will end up throwing an IllegalStateException because disable() was already called?

I guess the code that calls the cleanup action is guarding against exceptions, so this isn't a problem?

Reply:
Exactly - I'm relying on the fact that exceptions during cleaner actions are ok and ignored.